builtin/credential/cert/path_login.go

```diff
@@ -134,28 +135,47 @@ func (b *backend) verifyCredentials(req *logical.Request) (*ParsedCert, *logical
 	connState := req.Connection.ConnState
 
 	// Load the trusted certificates
-	roots, trusted := b.loadTrustedCerts(req.Storage)
+	roots, trusted, trustedNonCAs := b.loadTrustedCerts(req.Storage)
 
-	// Validate the connection state is trusted
-	trustedChains, err := validateConnState(roots, connState)
-	if err != nil {
-		return nil, nil, err
-	}
+	// If trustedNonCAs is not empty it means that client had registered a non-CA cert
+	// with the backend.
+	if len(trustedNonCAs) != 0 {
+		// Match the trusted chain with the policy
+		return b.matchNonCAPolicy(connState.PeerCertificates[0], trustedNonCAs), nil, nil
+	} else {
```

Comment:
This shouldn't be an else. These should be evaluated independently. If there are trusted non-CAs, see if there are any matches. If not, continue on with checking the CA-based roots.